Add python bindings for mesh and regular grid interpolant

nikwit opened the pull request:
## Proposed changes

added python bindings for Mesh and RegularGridInterpolant. This is to write a script that can interpolate an h5 file in post-processing.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nikwit committed:
Add python bindings for mesh

nikwit committed:
Add python bindings for RegularGridInterpolant

nilsvu commented on src/NumericalAlgorithms/Interpolation/Python/Bindings.cpp:
```diff
@@ -0,0 +1,17 @@
+
+namespace py = pybind11;
+
+namespace py_bindings {
```
Please wrap in the `intrp` namespace: `namespace intrp::py_bindings`

nilsvu commented on src/NumericalAlgorithms/Interpolation/Python/Bindings.cpp:
```diff
@@ -0,0 +1,17 @@
+
+namespace py_bindings {
+template <size_t Dim>
+void bind_regular_grid(py::module& m); //NOLINT
```
run clang-format over this (better: run `git-clang-format upstream/develop`)

nilsvu commented on src/NumericalAlgorithms/Interpolation/Python/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+
+set(LIBRARY "PyInterpolation")
+
+
```
delete 1 empty line

nilsvu commented on src/NumericalAlgorithms/Interpolation/Python/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+  ${LIBRARY}
+  PRIVATE
+  Interpolation
+  pybind11::module
```
also link `DataStructures` and `Spectral`

nilsvu commented on src/NumericalAlgorithms/Interpolation/Python/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+  PRIVATE
+  Interpolation
+  pybind11::module
+)
```
Add `spectre_python_add_dependencies(${LIBRARY} PyDataStructures PySpectral)`.

As we discussed, we'll probably have to move pybindings-dependencies to the `spectre_python_add_module` function, but that doesn't need to happen in this PR.

nilsvu commented on src/NumericalAlgorithms/Interpolation/Python/RegularGridInterpolant.cpp:
```diff
@@ -0,0 +1,33 @@
+  // the bindings are not complete and can be extended
+  py::class_<intrp::RegularGrid<Dim>>(
+      m, ("RegularGrid" + std::to_string(Dim) + "D").c_str())
+      .def(py::init([](const Mesh<Dim>& m1, const Mesh<Dim>& m2) {
```
- I don't think you need the lambda, this should be enough: `py::init<const Mesh<Dim>&, const Mesh<Dim>&>()`
- Add argument names for Python. They should be the same as in the C++ code, i.e. `source_mesh` and `target_mesh` in this case. See `src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Python/Tov.cpp` for an example.
- You can try to add the `const std::array<DataVector, Dim>&` optional constructor argument as well. The example linked above show you how to give it a default value. It might just work, but if you run into trouble don't worry about it.

nilsvu commented on src/NumericalAlgorithms/Interpolation/Python/RegularGridInterpolant.cpp:
```diff
@@ -0,0 +1,33 @@
+}
+template void bind_regular_grid<1>(py::module& m);
+template void bind_regular_grid<2>(py::module& m);
+template void bind_regular_grid<3>(py::module& m);
```
Might be nicer to define a non-template function that does these three calls here, so you don't have to template-instantiate them. See src/Domain/Python/ElementId.cpp for an example. Up to you.

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Bindings.cpp:
```diff
@@ -0,0 +1,22 @@
+void bind_basis(py::module& m); //NOLINT
+void bind_quadrature(py::module& m); //NOLINT
+template <size_t>
+void bind_mesh(py::module& m);  // NOLINT
```
same comment about defining a non-template function that calls a template detail function three times

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+
+set(LIBRARY "PySpectral")
+
+
```
delete 1 empty line

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Mesh.cpp:
```diff
@@ -0,0 +1,74 @@
+      .value("GaussLobatto", Spectral::Quadrature::GaussLobatto)
+      .value("CellCentered", Spectral::Quadrature::CellCentered)
+      .value("FaceCentered", Spectral::Quadrature::FaceCentered);
+}
```
move `bind_basis` and `bind_quadrature` to a separate file please. Perhaps `Spectral/Python/Spectral.cpp` because they are currently defined in `Spectral.hpp`.

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Mesh.cpp:
```diff
@@ -0,0 +1,74 @@
+
+namespace py = pybind11;
+
+namespace py_bindings {
```
wrap in `Spectral` namespace

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Mesh.cpp:
```diff
@@ -0,0 +1,74 @@
+template <size_t Dim>
+void bind_mesh(py::module& m) {  // NOLINT
+  //the bindings here are not complete
+  py::class_<Mesh<Dim>>(m, ("Mesh" + std::to_string(Dim) + "D").c_str())
```
Please add `py::arg` names to all function arguments below

nilsvu replied:
I think it would be useful to add a property `dim` to the Python class which is 1, 2 or 3 and static if possible. Would that be easy enough to add?

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/CMakeLists.txt:
```diff
@@ -44,3 +44,9 @@ add_dependencies(
   )
+
+spectre_add_python_bindings_test(
+  "Unit.Interpolation.Python.RegularGridInterpolator"
```
`RegularGrid` or `RegularGridInterpolant`

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/CMakeLists.txt:
```diff
@@ -44,3 +44,9 @@ add_dependencies(
   module_GlobalCache
   )
+
+spectre_add_python_bindings_test(
```
Please make a `Python/` subdirectory and move these binding tests there. (some older pybindings don't do that, but should)

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,148 @@
+    def polynomial1D(self, points):
+        return points**2 + points + 1
+
+    def polynomial2D(self, points):
```
it's a bit confusing what `points` means I think. How about you name it `coords` and write in the line below:

```py
x, y = coords
```

Then you can use `x` and `y` in the math expression

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,148 @@
+
+    def polynomial2D(self, points):
+        result = np.empty(len(points))
+        for i, p in enumerate(points):
```
<s>If I understand correctly this loop always iterates just the 2 dimensions. I'd prefer you just write the `2` instead of `len(points)` and explicitly fill `results[0] = ...` and `results[1] = ...`. Then probably also use two slightly different polynomials for the two dimensions.</s>

(See comment below) I suggest you just return a scalar (over the grid) in these polynomial functions, i.e. an array of shape `(num_points,)`

Same comment for the 3d case of course

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,148 @@
+                        p[2] + p[0] + 2
+        return result
+
+    def generate_Gauss_nodes(self, num_points):
```
Please make all function names in Python `snake_case`, i.e. `GaussLobatto` -> `gauss_lobatto`, `polynomial1D` -> `polynomial_1d` etc

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,148 @@
+
+    def test_Gauss_1D(self):
+        for num_points in range(3, 10):
+            m1 = Spectral.Mesh1D(num_points, Spectral.Basis.Legendre,
```
Please name these `source_mesh` and `target_mesh`

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,148 @@
+
+    #random polynomial functions of low order for exact interpolation
+    def polynomial1D(self, points):
+        return points**2 + points + 1
```
<s>I think it's easier to make this take and return a 2D array with only 1 row, so it works like the higher-dim functions</s>

(See comment below) Since I suggested to make the higher-dim functions return a scalar over the grid, this should also.

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,148 @@
+        initial_data = polynomial(initial_points)
+        target_data = polynomial(target_points)
+        interpolated_data = interpolant.interpolate(
+            DataStructures.DataVector(initial_data))
```
I don't quite understand why this works, or does it actually? If I understand correctly `initial_data` is a dim-dimension array, at least that's what the `polynomial` function returns, but `DataVector` expects a flattened array. I suggest you make clearer what the shapes of arrays are here. How about:
- Always work with `source_coords` and `target_coords` that are shaped `(dim, num_points)`, i.e. each component is flattened over the grid, but the components are not flattened.
- The polynomials return scalars, i.e. of shape `(num_points,)`. (I don't really understand how their components are currently handled)
- Name `generate_mesh` -> `logical_coordinates(mesh)` and put all the logic to generate the `(dim, num_points)`-shaped coords into that function. Then you don't need to have an extra test case per quadrature case below, but you can loop over the enum cases. If you prefer, you could also bind the `logical_coordinates(mesh)` function in `src/Domain/LogicalCoordinates.hpp`, which does the same thing. It returns a `tnsr::I` in C++, but you could make it return a numpy array in Python instead.

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,148 @@
+    def polynomial3D(self, points):
+        result = np.empty(len(points))
+        for i, p in enumerate(points):
+            result[i] = 3 * p[0] ** 2 + 2 * p[1] ** 2 + p[2] ** 2 + 2 * p[1] + \
```
You can break math (or just about anything) over multiple lines in Python by wrapping the expression in parens `()`

nilsvu commented on tests/Unit/NumericalAlgorithms/Spectral/CMakeLists.txt:
```diff
@@ -31,3 +31,9 @@ add_test_library(
   "Boost::boost;Domain;Spectral;SpectralHelpers"
   )
+
+spectre_add_python_bindings_test(
```
Also put in a `Python/` subdirectory please

nilsvu commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Mesh.py:
```diff
@@ -0,0 +1,104 @@
+                for i in range(100):
+                    extents1D = random.choices(self.extents, k=1)
+                    extents2D = random.choices(self.extents, k=2)
+                    extents3D = random.choices(self.extents, k=3)
```
this is cool 😄

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Mesh.cpp:
```diff
@@ -1,52 +1,39 @@
 // Distributed under the MIT License.
 // See LICENSE.txt for details.
 
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
```
Did you move this include to the top to mimic header-source pairs that include the header at the top? I have no opinion on that, might be a good idea. @nilsdeppe any strong stylistic opinion? :)

If you keep the header at the top, please add an empty line below.

You'll need to keep the `Spectral.hpp` include for the `Basis` and `Quadrature` types (don't delete it).

nilsdeppe replied:
The header order says this should _not_ be at the top. It's not the hpp file that corresponds/forward declares the contents of this cpp file, and therefore should be in with the others.

```
(If a cpp file with a corresponding hpp file:) hpp corresponding to cpp, followed by one blank line
```
first part is not true

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Mesh.cpp:
```diff
@@ -1,52 +1,39 @@
+  // the bindings here are not complete
   py::class_<Mesh<Dim>>(m, ("Mesh" + std::to_string(Dim) + "D").c_str())
+      .def_property_readonly_static("dim",
+                                    [](py::object) { return Dim; })  // NOLINT
```
Add `/*self*/` as the argument name, you may not need to `NOLINT` this anymore then.

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Spectral.cpp:
```diff
@@ -0,0 +1,29 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
```
Add empty line below

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Spectral.cpp:
```diff
@@ -0,0 +1,29 @@
+namespace Spectral::py_bindings {
+
+void bind_basis(py::module& m) {  // NOLINT
+
```
delete empty line

nilsvu commented on src/NumericalAlgorithms/Spectral/Python/Spectral.cpp:
```diff
@@ -0,0 +1,29 @@
+}
+
+void bind_quadrature(py::module& m) {  // NOLINT
+
```
delete empty line

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Python/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,135 @@
+            raise ValueError(
+                "Only Gauss or GaussLobatto are implemented quadratures")
+
+        mesh = np.meshgrid(*tuple(nodes for _ in range(dim)))
```
`np.meshgrid(*(dim * (nodes,)))` should also work. Up to you.

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Python/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,135 @@
+                source_coords = self.logical_coordinates(
+                    num_points, 1, quadrature)
+                target_coords = self.logical_coordinates(
+                    num_points + 2, 1, quadrature)
```
At this point you can make `logical_coordinates` just take the `mesh`. Better than essentially expanding the mesh data into the function :)

Then you need to do only these two things to merge the 1d, 2d and 3d functions into one:
1. Merge the `polynomial_*d` functions: Name `test_polynomial` or similar and determine the dimension from the input shape.
2. Get the `Mesh*D` class like this: `Mesh = [Mesh1D, Mesh2D, Mesh3D][dim - 1]`

We might be able to add a factory like `Mesh(dim=1, ...)`, though I'm not sure if that's necessary and it doesn't need to happen in this PR. That said, if you have an idea how to do that, feel free to give it a shot :D

nilsvu reviewed (changes requested):
The fixup looks good, please squash that in.

Recommendation: It helps to split fixup commits into the commits you want to squash them into, e.g. a `fixup Mesh` a `fixup RegularGrid` commit. Then you don't have to `git reset HEAD~` and split the changes when you want to squash.

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Python/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,86 @@
+
+    # random polynomial functions of low order for exact interpolation
+    def polynomial(self, coords):
+        if np.shape(coords)[0] == 1:
```
(Optional) You can just do `dim = len(coords)` and then switch on `dim`

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Python/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,86 @@
+    # random polynomial functions of low order for exact interpolation
+    def polynomial(self, coords):
+        if np.shape(coords)[0] == 1:
+            x = coords
```
(Optional) `x = coords[0]` so you don't need to `squeeze`

nilsvu commented on tests/Unit/NumericalAlgorithms/Interpolation/Python/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,86 @@
+
+                    source_coords = self.logical_coordinates(source_mesh)
+                    target_coords = self.logical_coordinates(target_mesh)
+                    print(np.shape(source_coords))
```
Delete this line

nilsvu reviewed (changes requested):
LGTM, only a few minor things, squash right away. Might as well do the two optional things when you delete the print that I guess is left-over from debugging. Up to you.

nilsvu commented:
@nikwit Would you please remove the "Feature/" from the PR title? Note that PR titles are listed in release notes, so it's good to format them consistently.

fmahebert commented on tests/Unit/NumericalAlgorithms/Interpolation/Python/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,89 @@
+
+class TestRegularGrid(unittest.TestCase):
+
+    # random polynomial functions of low order for exact interpolation
```
I suggest "random" -> "arbitrary", because it's not random in the sense of being randomly generated each time.

fmahebert commented on tests/Unit/NumericalAlgorithms/Interpolation/Python/Test_RegularGridInterpolant.py:
```diff
@@ -0,0 +1,89 @@
+
+        elif dim == 3:
+            x, y, z = coords
+            return 3 * x**2 + 2 * y**2 + z**2 + 2 * y + z + x + 2
```
[stylistic:] I think we prefer to write floats as "3.0" or "3." but not "3"... this applies to the prefactors on the various terms (but not the exponents which are integers). Same for 1D and 2D lines above.

fmahebert commented on tests/Unit/DataStructures/Tensor/Test_TensorData.py:
```diff
@@ -63,6 +63,7 @@ def test_extents_and_tensor_volume_data(self):
             "),(tensor component one, (1.5,1.1))))")
 
         # Tests for ExtentsAndTensorVolumeData functions
+
```
Should this blank line go above the comment? It has separated the comment from the function... also I think the comment's indentation level needs fixing.

fmahebert reviewed (commented):
I'm no expert on Python bindings, but overall this looks good to me. I have just three small comments that you can squash in directly.

nikwit commented:
@fmahebert thanks for the comments, I squashed them in. @nilsleiffischer I changed the name according to https://docs.github.com/en/github/administering-a-repository/renaming-a-branch which has been a feature of github for less than 2 weeks. As far as I can tell it was supposed to rename the branch and re-target all PRs, however it appears it just deleted the old branch and pushed a new one, automatically closing this PR. Not sure if its still buggy or I messed up.

It seems that I can now either restore the branch and re-open this PR with the old name or push one with the new name which will get rid of the review history. Which do you think is best?

nilsvu commented:
@nikwit Please restore the branch and re-open the PR. You can name the branch however you like, my comment referred to the _PR title_. You can just edit it at the top right of this page without doing git things 😄

nikwit commented:
riiight that makes more sense
